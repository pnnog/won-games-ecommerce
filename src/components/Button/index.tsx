import * as S from './styles';

export type ButtonProps = {
  children?: React.ReactNode;
  icon?: JSX.Element;
  size?: 'small' | 'medium' | 'large';
  fullWidth?: boolean;
};
const Button = ({
  children,
  icon,
  size = 'medium',
  fullWidth = false,
}: ButtonProps) => (
  <S.Wrapper size={size} fullWidth={fullWidth} hasIcon={!!icon}>
    {icon}
    {!!children && <span> {children} </span>}
  </S.Wrapper>
);

export default Button;
