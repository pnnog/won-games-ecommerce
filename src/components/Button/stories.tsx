import { Meta, Story } from '@storybook/react';

import * as icon from '@styled-icons/material-outlined';

import Button from '.';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string',
    },
    icon: {
      table: {
        disable: true,
      },
    },
  },
} as Meta;

export const Default: Story = (args) => <Button {...args} />;

Default.args = {
  children: 'Buy now',
};

export const withIcon: Story = (args) => <Button {...args} />;
withIcon.args = {
  size: 'small',
  children: 'Buy now',
  icon: <icon.AddShoppingCart />,
};
