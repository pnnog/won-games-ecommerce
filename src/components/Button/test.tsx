import { screen } from '@testing-library/react';

import { renderWithTheme } from '../../utils/tests/helpers';

import * as icon from '@styled-icons/material-outlined';

import Button from '.';

describe('<Button />', () => {
  it('should render the medium size byDefault', () => {
    const { container } = renderWithTheme(
      <Button> Children test render </Button>
    );

    expect(
      screen.getByRole('button', {
        name: /Children test render/i,
      })
    ).toHaveStyleRule('height', '4rem');
    expect(container).toMatchSnapshot();
  });

  it('should render small button if size small', () => {
    const { container } = renderWithTheme(
      <Button size="small"> Children test render </Button>
    );
    expect(
      screen.getByRole('button', {
        name: /Children test render/i,
      })
    ).toHaveStyle({ height: '3rem', padding: '0.8rem', 'font-size': '1.2rem' });
    expect(container).toMatchSnapshot();
  });

  it('should render large button if size large', () => {
    const { container } = renderWithTheme(
      <Button size="large"> Children test render </Button>
    );
    expect(
      screen.getByRole('button', {
        name: /Children test render/i,
      })
    ).toHaveStyleRule('height', '5rem');
    expect(container).toMatchSnapshot();
  });

  it('should render full width version', () => {
    const { container } = renderWithTheme(
      <Button fullWidth> Children test render </Button>
    );
    expect(
      screen.getByRole('button', {
        name: /Children test render/i,
      })
    ).toHaveStyleRule('width', '100%');
    expect(container).toMatchSnapshot();
  });

  it('should render a icon version', () => {
    const { container } = renderWithTheme(
      <Button icon={<icon.AddShoppingCart data-testid="icon" />}>
        Buy now test
      </Button>
    );

    expect(screen.getByTestId('icon')).toBeInTheDocument();
    expect(screen.getByText(/buy now test/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        This button is an link element
      </Button>
    );

    const element = screen.getByRole('link', {
      name: /This button is an link element/i,
    });
    expect(element).toHaveAttribute('href', '/link');
  });

  it('should render Button as a button when not is passed', () => {
    renderWithTheme(<Button>This button is an button element</Button>);

    const element = screen.getByRole('button', {
      name: /This button is an button element/i,
    });

    expect(element).toBeInTheDocument();
  });

  it('should render Button with minimal styles', () => {
    renderWithTheme(<Button minimal>Inscreva-se</Button>);

    const element = screen.getByRole('button', {
      name: /Inscreva-se/i,
    });

    expect(element).toHaveStyleRule('background', 'none');
    expect(element).toHaveStyleRule('background', 'none', {
      modifier: ':hover',
    });
  });
});
