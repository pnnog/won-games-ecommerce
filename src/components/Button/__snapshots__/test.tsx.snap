// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a icon version 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

<div>
   
  <button
    class="c0"
  >
    <svg
      aria-hidden="true"
      class="c1"
      data-testid="icon"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
      />
    </svg>
    <span>
       
      Buy now test
       
    </span>
  </button>
   
</div>
`;

exports[`<Button /> should render full width version 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  width: 100%;
}

<div>
   
  <button
    class="c0"
  >
    <span>
       
       Children test render 
       
    </span>
  </button>
   
</div>
`;

exports[`<Button /> should render large button if size large 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

<div>
   
  <button
    class="c0"
  >
    <span>
       
       Children test render 
       
    </span>
  </button>
   
</div>
`;

exports[`<Button /> should render small button if size small 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 3rem;
  font-size: 1.2rem;
}

<div>
   
  <button
    class="c0"
  >
    <span>
       
       Children test render 
       
    </span>
  </button>
   
</div>
`;

exports[`<Button /> should render the medium size byDefault 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<div>
   
  <button
    class="c0"
  >
    <span>
       
       Children test render 
       
    </span>
  </button>
   
</div>
`;
